### ISSUE TITLE
Feature request: LazyBatchNorm

### ISSUE THREAD

**Issue by imos** [pfnet/pytorch-pfn-extras#48]
In my project, I implemented LazyBatchNorm like this and I think pytorch-pfn-extras should have the similar feature:

```
class LazyBatchNorm1d(
    ppe.nn.modules.lazy.LazyInitializationMixin, torch.nn.BatchNorm1d
):
    """
    LazyBatchNorm1d is a lazy version of BatchNorm1d.  It does not require to
    specify the number of the input parameters beforehand.
    """

    lazy_parameter_names = ("weight", "bias")

    def __init__(self, num_features, *args, **kwargs):
        super().__init__(num_features or 0, *args, **kwargs)
        if num_features is None:
            self.num_features = None
            self.weight = ppe.nn.modules.lazy.UninitializedParameter()
            self.bias = ppe.nn.modules.lazy.UninitializedParameter()

    def forward(self, input):
        if isinstance(self.weight, ppe.nn.modules.lazy.UninitializedParameter):
            self.num_features = input.shape[-1]
            if self.affine:
                self.weight = torch.nn.Parameter(
                    self.weight.new_empty(self.num_features)
                )
                self.bias = torch.nn.Parameter(self.weight.new_empty(self.num_features))
            if self.track_running_stats:
                self.running_mean = torch.zeros(
                    self.num_features, device=self.running_mean.device
                )
                self.running_var = torch.ones(
                    self.num_features, device=self.running_var.device
                )
            self.reset_parameters()
        return super().forward(input)

    def reset_parameters(self):
        if self.lazy_parmeters_determined:
            super().reset_parameters()
```

Another discussion:
It would be good to have LazyBatchNorm instead of having LazyBatchNorm*d-s since lazy ones know its desired shapes. Any thoughts?

**Comment by kmaehashi** [comment]
Thank you for your contirbution!

> It would be good to have LazyBatchNorm instead of having LazyBatchNorm*d-s since lazy ones know its desired shapes. Any thoughts?

I feel explicitness is one of the PyTorch's philosophy because they provide `BatchNormXd` (X=1,2,3) even though everything is implemented in hidden `_BatchNorm` class. https://pytorch.org/docs/master/_modules/torch/nn/modules/batchnorm.html
I think it's better to do as the Romans do.

Separate `BatchNormXd` implementation will also make users easier to migrate from PPE's lazy implementation to PyTorch's upstream lazy implementation, which is under proposal in https://github.com/pytorch/pytorch/pull/40493

**Comment by emcastillo** [comment]
I will take care of this
